### PR TITLE
[core][distributed] simplify code to support pipeline parallel

### PR DESCRIPTION
**Pull request by youkaichao**
to minimize the line of code change for a model to support pipeline parallel.

in the model weight loading part, just add:

```python
                if name not in params_dict:
                    # in pipeline parallelism, we may have layers that are not
                    # present on this rank
                    continue
```

(the benefit is we don't need to introduce extra indentation of the try-except code)

in the layer construction part, just add:

```python
        self.start_layer, self.end_layer, self.layers = make_layers(
            config.num_hidden_layers,
            lambda: LlamaDecoderLayer(config=config,
                                      cache_config=cache_config,
                                      quant_config=quant_config))
```

hopefully, with these change, code for a model to support pp will be easier, and the review can also be easier.

### CODE REVIEW

**Comment by DarkLight1337** [comment]
Perhaps it would be better to move this function into the utils file inside `models/`?

**Comment by youkaichao** [comment]
> Perhaps it would be better to move the `make_layers` function into the utils file inside `models/`?

fixed in [347399e](https://github.com/vllm-project/vllm/pull/6406/commits/347399e2e62693f850fffb130565caccedcab7df)

File: vllm/model_executor/models/llama.py [hunk 1]
```diff
@@ -455,12 +445,16 @@ def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
...
                 if name.endswith(".bias") and name not in params_dict:
                     continue
-                try:
-                    param = params_dict[name]
-                    weight_loader = param.weight_loader
-                    weight_loader(param, loaded_weight, shard_id)
-                except KeyError:
-                    pass
+
+                if name not in params_dict:
```
[the remark this answers is not in the document]
Comment (youkaichao) [reply vllm/model_executor/models/llama.py:450]: this is quite difficult, as `load_weights` does not have layer information.

[the remark this answers is not in the document]
Comment (youkaichao) [reply vllm/model_executor/models/llama.py]: come up with a workaround in [61fa242](https://github.com/vllm-project/vllm/pull/6406/commits/61fa24286bd7de5468775f8b2de9360f00991cf5). PTAL!

**Comment by andoorve** [comment]
I noticed CUDA out of memory error on the basic correctness tests here. Is it reproducible locally? I think this change shouldn't cause that so it's possible a flaky test?

**Comment by youkaichao** [comment]
@andoorve finally figured it out, it is because lru cache stores a reference of the model, and then fails the gc system :(

**Comment by youkaichao** [comment]
merge first to unblock the following models support.

File: vllm/model_executor/models/utils.py [hunk 2]
```diff
@@ -39,3 +41,57 @@ def merge_vision_embeddings(input_ids: torch.Tensor,
...
+            missing_layer_names.append(name)
+    _model_to_pp_missing_layer_names[model_id] = missing_layer_names
+
+    return missing_layer_names
+
+
+def is_pp_missing_parameter(name: str, model: torch.nn.Module) -> bool:
+    """Check if a parameter is missing in a pipeline parallel model."""
+    for missing_layer_name in get_pp_missing_layer_names(model):
+        if name.startswith(missing_layer_name):
```
Comment (yudian0504) [inline vllm/model_executor/models/utils.py:95]: :-)  "xx.11".startswith("xx.1")

Comment (youkaichao) [reply vllm/model_executor/models/utils.py:95]: sorry for the bug, and thanks for pointing it out so quickly! please take a look at https://github.com/vllm-project/vllm/pull/6446 .

**Pull request merged**